File: apps/backend/src/domain/scheduler.js
const DEFAULT_BUFFER_TARGET_SECONDS = Number(process.env.STREAM_BUFFER_TARGET_SECONDS ?? 45);
const DEFAULT_RESPONSE_WINDOW_SECONDS = Number(process.env.REQUEST_RESPONSE_WINDOW_SECONDS ?? 90);

export function decideNextSegment(input) {
  const bufferSeconds = numberOrZero(input.bufferSeconds);
  const estimatedGenerationSeconds = Math.max(1, numberOrZero(input.estimatedGenerationSeconds));
  const oldestRequestAgeSeconds = Math.max(0, numberOrZero(input.oldestRequestAgeSeconds));
  const pendingRequestCount = Math.max(0, Number(input.pendingRequestCount ?? 0));
  const bufferTargetSeconds = Math.max(1, Number(input.bufferTargetSeconds ?? DEFAULT_BUFFER_TARGET_SECONDS));
  const responseWindowSeconds = Math.max(1, Number(input.responseWindowSeconds ?? DEFAULT_RESPONSE_WINDOW_SECONDS));

  const underrunRisk = estimatedGenerationSeconds / Math.max(1, bufferSeconds);
  const bufferDeficit = Math.max(0, bufferTargetSeconds - bufferSeconds) / bufferTargetSeconds;
  const requestPressure = pendingRequestCount === 0
    ? 0
    : Math.min(1, oldestRequestAgeSeconds / responseWindowSeconds);

  if (pendingRequestCount > 0 && requestPressure >= 0.85 && underrunRisk < 0.9) {
    return decision('serve_request', 'oldest request is close to the response target and buffer risk is acceptable', {
      priority: score(requestPressure + (1 - underrunRisk)),
      underrunRisk,
      bufferDeficit,
      requestPressure
    });
  }

  if (bufferSeconds < estimatedGenerationSeconds || bufferDeficit > requestPressure) {
    return decision('build_buffer', 'continuity buffer needs protection before reacting to more requests', {
      priority: score(bufferDeficit + underrunRisk),
      underrunRisk,
      bufferDeficit,
      requestPressure
    });
  }

  if (pendingRequestCount > 0) {
    return decision('serve_request', 'buffer is stable enough to respond to the request queue', {
      priority: score(requestPressure + 0.5),
      underrunRisk,
      bufferDeficit,
      requestPressure
    });
  }

  return decision('idle', 'no pending requests and buffer is within the target range', {
    priority: 0,
    underrunRisk,
    bufferDeficit,
    requestPressure
  });
}

function decision(action, reason, metrics) {
  return {
    action,
    reason,
    metrics: {
      priority: metrics.priority,
      underrunRisk: score(metrics.underrunRisk),
      bufferDeficit: score(metrics.bufferDeficit),
      requestPressure: score(metrics.requestPressure)
    }
  };
}

function numberOrZero(value) {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : 0;
}

function score(value) {
  return Math.round(Math.max(0, Math.min(1, value)) * 100) / 100;
}

File: apps/backend/app/scheduler.py
"""Port of the prototype buffer policy; does not generate or play audio."""

from math import floor


def score(value: float) -> float:
    return floor(max(0, min(1, value)) * 100 + 0.5) / 100


def decide_next_segment(*, buffer_seconds, estimated_generation_seconds,
                        oldest_request_age_seconds, pending_request_count,
                        buffer_target_seconds, response_window_seconds):
    estimated_generation_seconds = max(1, estimated_generation_seconds)
    risk = estimated_generation_seconds / max(1, buffer_seconds)
    deficit = max(0, buffer_target_seconds - buffer_seconds) / buffer_target_seconds
    pressure = min(1, oldest_request_age_seconds / response_window_seconds) if pending_request_count else 0
    if pending_request_count and pressure >= 0.85 and risk < 0.9:
        action = "serve_request"
        reason = "oldest request is close to the response target and buffer risk is acceptable"
        priority = pressure + 1 - risk
    elif buffer_seconds < estimated_generation_seconds or deficit > pressure:
        action = "build_buffer"
        reason = "continuity buffer needs protection before reacting to more requests"
        priority = deficit + risk
    elif pending_request_count:
        action = "serve_request"
        reason = "buffer is stable enough to respond to the request queue"
        priority = pressure + 0.5
    else:
        action = "idle"
        reason = "no pending requests and buffer is within the target range"
        priority = 0
    return {"action": action, "reason": reason, "metrics": {
        "priority": score(priority), "underrunRisk": score(risk),
        "bufferDeficit": score(deficit), "requestPressure": score(pressure),
    }}
